File: _08_02_/src/main/java/banco/dominio/ContaCorrente.java
package banco.dominio;

import banco.dominio.excecoes.ExcecaoChequeEspecial;

public class ContaCorrente extends Conta{
    private double chequeEspecial;

    public ContaCorrente(double saldoInicial) { this(saldoInicial,0); }

    public ContaCorrente(double saldoInicial, double chequeEspecial) {
        super(saldoInicial);
        this.chequeEspecial = chequeEspecial;
    }

    @Override
    public void sacar(double valor) throws ExcecaoChequeEspecial {
        // Saldo + valor cheque especial suficientes para cobrir o saque
        try {
            double deficit = this.saldo - valor;
            if (deficit < 0) {
                if (this.chequeEspecial == 0D) {
                    throw new ExcecaoChequeEspecial(-deficit, "Nao ha cheque especial");
                } else if (this.chequeEspecial < -deficit) {
                    throw new ExcecaoChequeEspecial(-deficit, "Saldo insuficiente no cheque especial");
                }
            }

            this.saldo = deficit;
        } catch (ExcecaoChequeEspecial e) {
            throw e;
        }
    }

    public double getChequeEspecial() { return chequeEspecial; }

    public void setChequeEspecial(double chequeEspecial) { this.chequeEspecial = chequeEspecial; }
}

File: Trabalho_JS/banco/dominio/ContaCorrente.js
import ExcecaoChequeEspecial from './excecoes/ExcecaoChequeEspecial.js';
import Conta from './Conta.js';

export default class ContaCorrente extends Conta {
    #chequeEspecial;

    constructor(saldoInicial, chequeEspecial) {
        super(saldoInicial);
        this.#chequeEspecial = chequeEspecial;
    }

    sacar(valor) {
        if (this._saldo + this.#chequeEspecial >= valor) {
            this._saldo -= valor;
         } else {
            let deficit = valor - this._saldo;
            if (this.#chequeEspecial == 0) {
                throw new ExcecaoChequeEspecial(deficit, "Não há cheque especial");
            } else {
                throw new ExcecaoChequeEspecial(deficit, "Saldo insuficiente no cheque especial");
            }
        }
    }

    getChequeEspecial() { return this.#chequeEspecial; }

    setChequeEspecial(chequeEspecial) { this.#chequeEspecial = chequeEspecial; }
}